Config UI: add missing config sections / partial yaml inputs (BC)

naltatis opened the pull request:
Replaces https://github.com/evcc-io/evcc/pull/11739

**Goal**
- 🐭 make smaller config chunks (tariff, messages, mqtt, ...) editable via UI instead of complete evcc.yaml.
- 🚴‍♀️ create UI and API (endpoints, persistance and initialization) for all missing config areas in a **fast and pragmatic** way.

**Later steps (seperate PRs)**
- 🎩 Improve UX where necessary (tariff template UI #10042, ...)
- 📟 Allow yaml-snippets as an alternative UI config option for vehicles, meters, chargers (e.g. to allow plugins)

**BC**

- all /api/state durations are now in seconds instead of nanoseconds

**Todos**

- [x] api structure for all missing config elements @naltatis 
- [x] port yaml editor from https://github.com/evcc-io/evcc/pull/11739 @naltatis 
- [x] mqtt UI @naltatis 
- [x] mqtt API @andig 
- [x] Influx UI @naltatis 
- [x] Influx API @andig 
- [x] eebus UI @naltatis
- [x] eebus API @andig 
- [x] tariffs UI @naltatis
- [x] tariffs API @andig 
- [x] messaging UI @naltatis
- [x] messaging API @andig 
- [x] modbusproxy UI @naltatis
- [x] modbuxproxy API @andig 
- [x] interval UI @naltatis
- [x] interval API @andig (@naltatis time.Duration oder einfach in Sekunden?)
- [x] sponsortoken API @andig 
- [x] network UI @naltatis
- [x] network API @andig 
- [x] maxgridsupply UI @naltatis
- [x] maxgridsupply API @andig 
- [x] circuit UI @naltatis 
- [x] config status endpoint / decision needed ([see here](https://github.com/evcc-io/evcc/pull/13319/files#diff-67960f885830ef6514bf0bd84bea170135f8a3dee2857aa8f1ba7fa7f87afa4bR163-R166)) @naltatis @andig 
- [x] ~remove all ns conversions from UI~ @naltatis 
- [x] fix messaging broken by encoded state values @andig 
- [x] fix broken date encoding (see https://github.com/evcc-io/evcc/pull/13319/commits/55b5cfd0879cf799350fb4d134de669dff1f302f) @andig 
- [x] persist residualPower @andig
- [x] decide messaging migration and differentiate services vs events @naltatis @andig 
- [x] decide tariff migration @naltatis @andig 
- [x] decide eebus migration @naltatis @andig 
- [x] decide hems migration @naltatis @andig 
- [x] decide modbusproxy migration @naltatis @andig 
- [x] circuit API @andig 
- [x] sponsortoken UI @naltatis
- [x] switch to api-fied sponsor struct for UI @naltatis (https://github.com/evcc-io/evcc/pull/13319/commits/721d35566f63aed59e76cb5c7f09de47f8d26898)
- [x] Config-Endpunkte auch bei Startup Error bereitstellen @andig @naltatis 
- [x] use error classes in UI to highlight faulty configuration @naltatis 
- [x] fix MQTT/Influx (others?) publishing passwrds
- [x] read interval db configuration on start @andig 
- [x] use `temp` in addition to `soc` @naltatis 
- [x] add migration & clean command (opt-in for testing) @andig 
- [x] e2e tests @naltatis 
- [x] improve main-UI + fatal error @naltatis 

Out of scope:

- [ ] ~move `dirty` api to web socket~
- [ ] ~move site meters to global config, identified by type (grid, pv, battery, aux)~
- [ ] ~show/update meter configuration when device fails during startup~
- [ ] ~allow saving broken devices (RCT)~

→ moved out-of-scope tasks to epic https://github.com/evcc-io/evcc/issues/6029

naltatis committed:
yaml editor basics

naltatis committed:
Merge branch 'master' into feature/config_partial_yaml

naltatis committed:
add structure for missing config endpoints

naltatis commented on server/http.go:
```diff
@@ -141,6 +141,27 @@ func (s *HTTPd) RegisterSiteHandlers(site site.API, cache *util.Cache) {
 		"deletedevice": {"DELETE", "/devices/{class:[a-z]+}/{id:[0-9.]+}", deleteDeviceHandler},
 		"testconfig":   {"POST", "/test/{class:[a-z]+}", testConfigHandler},
 		"testmerged":   {"POST", "/test/{class:[a-z]+}/merge/{id:[0-9.]+}", testConfigHandler},
+		// new endpoints ⤵︎
```
@andig ich hab hier mal alle aus meiner Sicht noch fehlenden Endpoints runtergeschrieben damit wir die evcc.yaml ersetzen können. Ums einfach zu halten jeweils mit ausmodelliertem GET/PUT Handler und passenden Payload Objekten. Bau das gerne so um, wie es am effizientesten ist. Aber auf der Basis dieser Endpunkte und Datentypen würde ich jetzt anfangen einfache UI-Masken zu bauen.

naltatis committed:
lint

naltatis committed:
Merge branch 'master' into feature/config_partial_yaml

naltatis committed:
toml

naltatis committed:
e2e

andig commented:
@naltatis ich brauche mal eine Idee. ModbusProxy ist eine Liste. Wie wollen wir hier mit Updates umgehen:
- per Element (über Index? über Port?) (json je Element)
- ganze Liste am Stück (json)?
- andere Ideen?
Wäre super einen Ansatz zu finden, der auch mit anderen Integrationen, z.B. den Push Service funktionieren kann.

Meine Tendenz wäre per Element über numerischen Index. Passt das?

naltatis commented:
@andig Im ersten Schritt (dieser PR) würde ich dir nur einen Yaml Block geben und keine Rücksicht auf die innere Struktur nehmen. Heißt, ein Update hat immer ein komplettes Neu-initialisieren zur Folge. Nichts auf Item-Ebene. Gleiches gilt für Messaging. Da ist die Struktur ja noch komplizierter, weil Messages und Services (Liste).

Das ist natürlich nicht schön und wir müssen das früher oder später sauber als Struktur/Liste ausmodellieren. Aber ich sehe das hier erstmal als einen pragmatischen Zwischenschritt.

andig commented:
Passt, das ist hässlich aber einfach.

andig committed:
wip to be reverted

naltatis committed:
Merge branch 'master' into feature/config_partial_yaml

deadrabbit87 committed:
Docs: remove sungrow interface limitations (#13381)

andig committed:
Add a bunch of settings

andig commented:
@naltatis die existierenden Site APIs, z.B: `buffersoc` haben alle keine Getter. Brauchen wir trotzdem `GET`APIs für die unstrukturierten Werte wie interval? Das erscheint inkonsistent.

andig committed:
wip

andig committed:
wip

VolkerK62 committed:
keba-modbus: additional info regarding phase switch (#13415)

andig committed:
chore: upgrade libraries

andig committed:
Log Viewer: Allow retrieving logs from ring buffer (#13330)

andig committed:
wip

andig committed:
Align encodings

andig committed:
Add Warp3 charger (#13422)

premultiply committed:
fronius-ohmpilot: swap power sign

Fixes #13405

andig committed:
Implement sponsors api

andig committed:
API-fy sponsor status

andig committed:
wip

andig committed:
wip

andig committed:
wip

ben-bole committed:
Update solaredge-hybrid.yaml (#13418)

andig committed:
chore: use errors.As where appropriate

TRON4R committed:
Fiat: require pin for updated soc (#13223)

andig committed:
Fiat: ignore HTTP 403 during deep refresh (#13428)

andig committed:
chore: don't retry permanent errors

andig committed:
chore: minor

andig committed:
chore: upgrade modbus

weblate committed:
Translations update from Hosted Weblate (#13379)

* Translated using Weblate (German)

Currently translated at 99.4% (386 of 388 strings)

Co-authored-by: Christoph Schmelter <hfmargarine@googlemail.com>
Translate-URL: https://hosted.weblate.org/projects/evcc/evcc/de/
Translation: evcc/evcc

* Translated using Weblate (Portuguese)

Currently translated at 100.0% (388 of 388 strings)

Co-authored-by: kfussmann <karlfussmann@gmx.de>
Translate-URL: https://hosted.weblate.org/projects/evcc/evcc/pt/
Translation: evcc/evcc

* Translated using Weblate (Swedish)

Currently translated at 99.2% (385 of 388 strings)

Co-authored-by: Jonas Gate <tidy.koala8317@brot.se>
Translate-URL: https://hosted.weblate.org/projects/evcc/evcc/sv/
Translation: evcc/evcc

* Translated using Weblate (Lithuanian)

Currently translated at 100.0% (409 of 409 strings)

Co-authored-by: RTTTC <94727758+RTTTC@users.noreply.github.com>
Translate-URL: https://hosted.weblate.org/projects/evcc/evcc/lt/
Translation: evcc/evcc

* fix toml

---------

Co-authored-by: Christoph Schmelter <hfmargarine@googlemail.com>
Co-authored-by: kfussmann <karlfussmann@gmx.de>
Co-authored-by: Jonas Gate <tidy.koala8317@brot.se>
Co-authored-by: RTTTC <94727758+RTTTC@users.noreply.github.com>
Co-authored-by: premultiply <4681172+premultiply@users.noreply.github.com>

andig committed:
Revert "Fiat: require pin for updated soc (#13223)"

This reverts commit 3740a31990f403369be1247f18b59802dd4f713d.

andig committed:
chore: simplify

VolkerK62 committed:
Typo: in UI notifications (#13438)

naltatis committed:
Log UI: fixed button layout (#13442)

andig committed:
Log viewer: fix old data returned when count set

andig committed:
chore: minor

andig committed:
chore: reduce memory and allocations

naltatis committed:
Log UI: minor improvements (#13453)

andig committed:
Battery control: limit at 100%

andig committed:
chore: minor

naltatis committed:
Merge branch 'master' into feature/config_partial_yaml

naltatis commented:
@andig nach dem Merge bauts bei mir nicht mehr. Woher kommt `EncodeAny`?

```
$ make build
...
# github.com/evcc-io/evcc/util
util/cache.go:66:21: undefined: EncodeAny
util/cache.go:73:20: undefined: EncodeAny
make: *** [build] Error 1
```

andig commented:
Im Master gibts das:

```
// EncodeAny provides a consumer-friendly default encoding for any type
func EncodeAny(v any) any {
```

naltatis commented:
@andig Wo? Ich hab überall gesucht, kann das aber nicht finden.

andig commented:
Siehe Mail, muss ich prüfen.

naltatis committed:
fix build

naltatis committed:
make test green; rework needed

naltatis committed:
remove sponsor status endpoint > socket state

naltatis committed:
ui lint

naltatis committed:
fix deadlock

naltatis committed:
fix e2e

naltatis commented:
> remove all ns conversions from UI

Let's move this unification/cleanup to a separate PR. Currently, all `Durations` over socket are in seconds (thats relevant for this PR). The durations in the REST endpoints for state and sessions are in nanoseconds, and for plan we have a REST endpoint that returns duration in seconds manually ([see here](https://github.com/evcc-io/evcc/blob/540ddee34abbb490948041f047b1be46bdac7afd/server/http_loadpoint_handler.go#L63-L64)).

naltatis committed:
sponsorship changes

naltatis commented on server/http_config_site_other_handler.go:
```diff
@@ -0,0 +1,26 @@
+
+	// TODO find better place
+	settings.SetString(keys.SponsorToken, token)
+	setConfigDirty()
```
@andig Hier muss auch ein Publishing des neuen Zustands (`sponsor`) passieren. Sonst hat die UI keine Chance dem User das Ergebnis seiner Aktion (verlängertes Token, anderer Name, ...) anzuzeigen.

Auch das Entfernen eines Tokens (POST Leestring, oder DELETE) müssen wir noch ermöglichen.

naltatis committed:
yaml edit dialog

andig commented:
Hab ich- das ist generell der Punkt "Verdrahtung" aus den Todos

naltatis committed:
new icons, modbusproxy, ...

naltatis committed:
refactor restart behaviour; wip

naltatis committed:
Merge branch 'master' into feature/config_partial_yaml

naltatis committed:
unignore yaml in js

naltatis committed:
refactor dirty ui

naltatis committed:
surpress initial auth/telemetry errors when server not available

andig committed:
Yaml: return line number on errors

naltatis committed:
hems ui; restart refactor

naltatis committed:
highlight yaml error line

naltatis commented on server/socket_test.go:
```diff
@@ -18,12 +18,12 @@ func TestEncode(t *testing.T) {
 		{"minpv", "\"minpv\""},
 		{time.Time{}, "null"},
-		{now, "\"" + now.Format(time.RFC3339) + "\""},
+		{now, "\"\\\"" + now.Format(time.RFC3339) + "\\\"\""},
```
@andig magst du dir Fehler noch mal anschauen. Ich hatte nur den Unit Test angepasst, damit der keine Fehler wirft. Ist aber eigentlich falsch. Wenn das Datum wieder korrekt ausgegeben wird laufen auch die Integrationstests hier im Branch wieder (Planner-Tests scheitern).

naltatis committed:
unified restart/offline handling; block input while offline/restarting

andig committed:
wip

andig committed:
Make encoder behavior configurable

andig committed:
wip

andig committed:
Merge branch 'master' into feature/config_partial_yaml

andig committed:
Remove ocpp proxy

andig committed:
Document api encoding

andig committed:
Return json settings

andig commented on util/sponsor/auth.go:
```diff
@@ -63,3 +72,25 @@ func ConfigureSponsorship(token string) error {
+
+// Status returns the sponsorship status
+func Status() StatusStruct {
+	expiresSoon := false
```
```suggestion
	var expiresSoon bool
```

andig committed:
Update util/sponsor/auth.go

naltatis committed:
saving structured data in a generalized way, modal improvements

naltatis commented:
**Screenshots of edit dialoges for structured data and gerneral yaml stuctures (temporary)**

![Bildschirmfoto 2024-04-21 um 23 09 09](https://github.com/evcc-io/evcc/assets/152287/1a3d61da-5ee8-4a66-be22-9572bc63f97d)

![Bildschirmfoto 2024-04-21 um 23 08 58](https://github.com/evcc-io/evcc/assets/152287/3d6d9b04-e5ba-4baf-9d6a-cb62d3454cc1)

![Bildschirmfoto 2024-04-21 um 23 08 38](https://github.com/evcc-io/evcc/assets/152287/b685244b-6a60-4991-a86d-cfeb74d135df)

![Bildschirmfoto 2024-04-21 um 23 08 33](https://github.com/evcc-io/evcc/assets/152287/2d88d464-9c05-44eb-bd5b-8db9e5fd2d34)

naltatis committed:
prevent double scrollbar for modal

andig committed:
Migrate mqtt, influxdb and eebus

andig committed:
Simplfy routing

naltatis commented on server/http.go:
```diff
@@ -135,6 +140,59 @@ func (s *HTTPd) RegisterSiteHandlers(site site.API, auth auth.Auth, cache *util.
+	} {
+		configRoutes[key] = route{Method: "GET", Pattern: "/" + key, HandlerFunc: settingsGetStringHandler(key)}
+		configRoutes["update"+key] = route{Method: "POST", Pattern: "/" + key, HandlerFunc: settingsSetYamlHandler(key, struc)}
+		configRoutes["delete"+key] = route{Method: "DELETE", Pattern: "/" + key, HandlerFunc: settingsDeleteHandler(key)}
```
```suggestion
```

Bei YAML Eigenschaften brauchen wir keinen DELETE Endpunkt. Löschen ist da eine leere oder auskommentierte Payload. Das funktioniert mit dem jetzigen POST auch schon. Abgesehen davon würde `settingsDeleteHandler` aktuell JSON in ein YAML Feld schreiben (`{}`).

naltatis commented on server/http.go:
```diff
@@ -135,6 +140,59 @@ func (s *HTTPd) RegisterSiteHandlers(site site.API, auth auth.Auth, cache *util.
+		// "deletenetwork":      {"DELETE", "/network", settingsDeleteHandler("network")},
+		"interval":           {"POST", "/interval/{value:[0-9.]+", settingsSetDurationHandler("interval")},
+		"updatesponsortoken": {"POST", "/sponsortoken/{token:[a-zA-Z0-9_-.]+}", updateSponsortokenHandler},
+		"deletesponsortoken": {"DELETE", "/sponsortoken", settingsDeleteHandler(keys.SponsorToken)},
```
Hier würd ich lieber nen dedizierten Endpunkt haben, der auch den aktualisierten Sponsor-Status korrekt neu publisht. Im Gegensatz zu den anderen Feldern find ich es hier sehr wichtig, dass der Nutzer eine direkte Bestätigung/Validierung in der UI bekommt. Also ohne Neustart.

andig replied:
Räums gerne auf- der letzte ist bewusst noch drin. Wir brauchen wohl 2 Delete Handler- empty und leeres JSON.

naltatis replied:
Ich hab den handler nun so umgebaut, dass delete immer nen leeren String setzt. Dafür kann der getJson handler damit jetzt umgehen und liefert den default struct zurück. vorher kam stattdessen immer ein leeres `{}` wenn einmal DELETEed wurde.

naltatis committed:
Merge branch 'master' into feature/config_partial_yaml

naltatis commented:
@andig Wir müssen noch das Problem lösen, das Auftritt, wenn ein Fehler in der Konfiguration dazu führt, dass evcc nicht starten kann (Startup Error). Aktuell sind die `/api/config` Endpunkte dann nicht erreichbar und der Nutzer kann den Fehler nicht korrigieren. Die analog zu `RegisterAuthHandlers` auszulagern und in beiden Zuständen (Normal vs. Startup Error) zu verwenden geht prinzipiell. Gibt aber zwei Probleme:

1. `GET /config/site` & `POST /config/site` brauchen eine aktive Site. Die haben wir in dem Zustand heute nicht. Diese APIs sind wichtig, weil die UI daran erkennt, welche Meter welche Funktion haben und das Entfernen eines Meters ohne nicht funktioniert.
2. Bei den "einfachen Endpunkten" (sponsor, interval, maxgridsupply, ...) nutzen wir WS als Datenrückkanal (lesen). Aktuell wird in "Startup Error" Modus nur `config/fatal/file` gepublisht, nicht die üblichen Werte.

Hast du eine Idee, wie wir diese Themen elegant und pragmatisch lösen können?

**Mittelfristige Lösung:** Wir sollten überlegen, wie wir diese harte Modus-Unterscheidung (Normal vs. Startup Error) loswerden können. Also beim Hochfahren immer eine Seite und alle API Endpunkte haben. Den Indikator, dass es fundamentale Probleme gibt, können wir ja trotzdem publishen und in der UI darauf reagieren. Aber dann eher als Fehlermeldung auf der Config/Main-Seite und nicht als eigene Fehlerseite.

naltatis committed:
return empty json struct if no config exists

naltatis committed:
pw manager annotation, required fields

naltatis committed:
move instance creation into handler to avoid inconsistancies between update/delete/get

andig commented on server/http_global_settings_handler.go:
```diff
@@ -87,30 +86,32 @@ func settingsSetYamlHandler(key string, struc any) http.HandlerFunc {
 }
 
-func settingsSetJsonHandler(key string, struc any) http.HandlerFunc {
+func settingsSetJsonHandler(key string, struc func() interface{}) http.HandlerFunc {
```
Die Änderung finde ich nicht schön. Warum ist das notwendig?

naltatis commented on api/globalconfig/types.go:
```diff
@@ -0,0 +1,108 @@
+	Org      string
+	User     string
+	Password string
+	Interval time.Duration
```
@andig ist das unused? Dann lass ichs in der UI einfach raus.

naltatis replied:
`Influx.Interval`?

naltatis replied:
(der Code-Ausschnitt ist etwas zu klein)

naltatis committed:
Add InfluxDB UI

naltatis committed:
Add eebus ui

naltatis committed:
add influx org/bucket; i18n

naltatis committed:
i18n

naltatis commented:
A few screenshots of the new dialoges:

![Bildschirmfoto 2024-04-24 um 17 03 36](https://github.com/evcc-io/evcc/assets/152287/3fc18610-8894-4b8f-b002-9e1ed511f424)

![Bildschirmfoto 2024-04-24 um 17 03 09](https://github.com/evcc-io/evcc/assets/152287/372df54e-2726-43ba-87f9-5de1ddefd6e0)

![Bildschirmfoto 2024-04-24 um 17 06 23](https://github.com/evcc-io/evcc/assets/152287/4ae383be-5b55-4bdd-bfcd-a11d2788fb90)

![Bildschirmfoto 2024-04-24 um 17 04 13](https://github.com/evcc-io/evcc/assets/152287/6e5615bc-3fb2-4e18-8afd-a0d1f10491a0)

![Bildschirmfoto 2024-04-24 um 17 08 51](https://github.com/evcc-io/evcc/assets/152287/5b206c7a-9a00-49f1-b847-31b777b5320f)


![Bildschirmfoto 2024-04-24 um 17 04 01](https://github.com/evcc-io/evcc/assets/152287/e7100887-c34d-46d9-93a7-c2702c8b97c2)

![Bildschirmfoto 2024-04-24 um 17 03 55](https://github.com/evcc-io/evcc/assets/152287/6041efaa-f075-40f9-bb23-62e2e4ccc68e)

![Bildschirmfoto 2024-04-24 um 17 04 06](https://github.com/evcc-io/evcc/assets/152287/0dc3a5c6-e290-4780-94eb-0893bf434480)

andig committed:
Cleanup

naltatis commented:
@andig magst du hier mal den Merge Konflikt mit dem neuen Circuit auflösen?

naltatis committed:
visual cleanup

naltatis commented:
Cleanere Kacheln (kein Border, keine Badges, Quartett-Karten-Style 🤓)

![Bildschirmfoto 2024-04-30 um 20 39 45](https://github.com/evcc-io/evcc/assets/152287/e9d2d992-262f-4661-87e8-35d487b13265)

andig committed:
Add class and device errors, split site config

andig committed:
Marshal error json

naltatis committed:
add tariff status

andig committed:
Merge branch 'master' into feature/config_partial_yaml

andig committed:
wip

andig committed:
Merge branch 'master' into feature/config_partial_yaml

andig commented:
@naltatis Mergekonflikte sind weg, TOML scheint er nicht zu mögen- fehlt da tatsächlich einfach die richtige Formatierung?

naltatis commented:
> @naltatis Mergekonflikte sind weg, TOML scheint er nicht zu mögen- fehlt da tatsächlich einfach die richtige Formatierung?

Ich schau

naltatis committed:
toml fmt

naltatis commented:
Jo, toml war noch inkorrekt. Ist mir vorher nicht aufgefallen weil die Pipeline nicht lief. Ist gefixt. Allerding schlagen auch noch tests fehl:

```
Error: cmd/setup_circuits_test.go:39:11: undefined: globalConfig
```

naltatis committed:
cleaner device status

andig committed:
wip

andig commented:
@naltatis integration failed wieder oder noch

naltatis committed:
fix test result test

naltatis commented:
> @naltatis integration failed wieder oder noch

War erwartungskonform. Der Grund war meine Änderung kurz davor.

naltatis committed:
Merge branch 'master' into feature/config_partial_yaml

naltatis committed:
tidy

naltatis committed:
Merge branch 'master' into feature/config_partial_yaml

naltatis committed:
added circuits ui; updated icons; general settings alignment; german translation

naltatis committed:
fix tests; generic title modal

naltatis committed:
toml

naltatis committed:
fix tests

naltatis commented on server/http_global_settings_handler.go:
```diff
@@ -0,0 +1,115 @@
+		}
+
+		settings.SetInt(key, int64(time.Second*time.Duration(val)))
+		setConfigDirty()
```
Die UI braucht bei Erfolg ein publish des neuen Werts. Sonst wird bis nach dem Neustart der alte Wert angezeigt und der Nutzer ist sich nicht sicher ob jetzt gespeichert wurde oder nicht. Analog zum Kommentar beim Sponsor Token.

naltatis commented:
@andig Die Antwort von `GET /api/config/network` liefert die Null Werte des Structs, aber nicht die eigentlichen Defaults zurück.

```
GET http://localhost:7071/api/config/network

{
  "result": {
    "Schema": "",
    "Host": "",
    "Port": 0
  }
}
```

Wurde noch keine Konfiguration vorgenommen, würde ich erwarten, dass ich die Defaults bekomme:
https://github.com/evcc-io/evcc/blob/feature/config_partial_yaml/cmd/setup.go#L59-L63

Gleiches gilt für das `topic: "evcc"` bei MQTT. Dafür hab ich den Default bislang noch mal in der UI parallel gepflegt. Schöner wäre aber, wenn wir das nicht doppeln müssten.
https://github.com/evcc-io/evcc/blob/feature/config_partial_yaml/assets/js/components/Config/JsonModal.vue#L105

naltatis committed:
added network ui

naltatis commented:
Network UI

<img width="596" alt="Bildschirmfoto 2024-05-11 um 16 39 29" src="https://github.com/evcc-io/evcc/assets/152287/f642be09-cf74-4d34-8d18-8aba62ce0194">

naltatis committed:
Merge branch 'master' into feature/config_partial_yaml

naltatis commented:
@andig bei `residualPower` fehlt noch die Persistenz.

naltatis committed:
control settings

naltatis commented:
Added `interval`, `residualPower`, `maxGridSupply` as **Control behavior** / **Regelverhalten** group.

<img width="594" alt="Bildschirmfoto 2024-05-14 um 21 53 44" src="https://github.com/evcc-io/evcc/assets/152287/711035d6-7367-4d3f-b709-3de3b6761305">

<img width="1342" alt="Bildschirmfoto 2024-05-14 um 21 54 01" src="https://github.com/evcc-io/evcc/assets/152287/63ae22f0-fb53-4c78-ae78-1c54bca4c33e">

naltatis committed:
translation; prevent scroll jumps for modal

naltatis committed:
sponsor modal WIP

VolkerK62 commented:
würde an manchen Stellen z.B. Batterieeinstellungen, Loadpoint threshold/delay ...  eine Art "reset to default" Sinn machen?

naltatis commented:
> würde an manchen Stellen z.B. Batterieeinstellungen, Loadpoint threshold/delay ... eine Art "reset to default" Sinn machen?

Ja definitiv! Würd ich aber in einer späteren Iteration sehen damit dieser PR nicht noch größer wird 😄

naltatis committed:
Merge branch 'master' into feature/config_partial_yaml

naltatis committed:
lint

andig committed:
Fix network, mqtt, influx

andig committed:
Fix json publishing- initial and on update

andig committed:
Convert EEbus to yaml

andig committed:
Migrate residualPower

andig committed:
Add messaging (wip)

andig committed:
Complete migrations

andig committed:
wip

andig committed:
Add more error classes

andig committed:
Stop validating config file errors and start UI on error

andig committed:
wip

andig committed:
Reorder handlers for splitting

andig committed:
Split config api from site and loadpoints

andig committed:
wip

andig committed:
Handle tariff errors

andig committed:
Make state globally available

andig committed:
Fix re-marshaling mapstructure:remain elements inline

naltatis committed:
fix log e2e

naltatis committed:
revert influx init

naltatis committed:
publish: lower case, omit empty and passwords

naltatis committed:
startup error

naltatis committed:
Merge branch 'master' into feature/config_partial_yaml

andig commented:
Sieht super aus! Mit der error class solltest du auch direkt mqtt rot highlighten können?

naltatis commented:
Wo finde ich die? Ich hab nur was in 'fatal' gesehen.

andig committed:
Structured fatal errors

andig commented:
Aber jetzt:

```json
{
  "result": {
    "auth": {
      "vehicles": {}
    },
    "fatal": {
      "class": "mqtt",
      "error": "failed configuring mqtt: error connecting: network Error : dial tcp: lookup foo.bar: no such host"
    },
    "influx": {},
    "interval": 3,
    "loadpoints": [],
    "mqtt": {
      "topic": "evcc"
    },
    "network": {
      "schema": "http",
      "host": "evcc.local",
      "port": 7070
    }
  }
}
```

andig committed:
wip

andig commented:
> bich musste den !== "" Check wieder reverten. Ohne den stoppt das Publishing und die UI bekommt keine Werte Updates. Dadurch sind auch die e2e Tests kaputt gegangen.

Den brauchen wir aber- sonst funktioniert die Migration des Settings nicht. Mir ist nicht klar, was da stoppen sollte? Evtl. vorherige Fehlmigration? Aber auch dann sollte nix stoppen!

naltatis commented:
> Evtl. vorherige Fehlmigration?

Ne, das passiert auch bei einer neuen Installation (siehe E2E tests). Du kannst das ja einfach reproduzieren.

andig commented:
Hab ich kein Problem. Welches konkrete Peoblem hast du denn?

andig commented:
@naltatis ich sehe kein Problem nach revert:

```
[site  ] DEBUG 2024/05/23 19:20:25 set battery mode: normal
[influx] ERROR 2024/05/23 19:20:26 write failed (attempts 1): Post "http://foo.bar/api/v2/write?bucket=bar&org=foo&precision=s": dial tcp: lookup foo.bar: no such host
[site  ] DEBUG 2024/05/23 19:20:28 ----
[lp-1  ] DEBUG 2024/05/23 19:20:28 charge power: 4600W
[lp-2  ] DEBUG 2024/05/23 19:20:28 charge power: 0W
[site  ] DEBUG 2024/05/23 19:20:28 pv power: 8088W
[site  ] DEBUG 2024/05/23 19:20:28 battery soc: 55%
[site  ] DEBUG 2024/05/23 19:20:28 battery power: 0W
[site  ] DEBUG 2024/05/23 19:20:28 grid meter: -2996W
[site  ] DEBUG 2024/05/23 19:20:28 site power: -2996W
[lp-2  ] DEBUG 2024/05/23 19:20:28 charger status: B

```

andig committed:
Re-apply influx

naltatis committed:
fatal error highlight; fix mqtt value publishing and startup

andig commented on provider/mqtt/client.go:
```diff
@@ -26,11 +26,11 @@ func ClientID() string {
-	Insecure bool   `json:"insecure,omitempty"`
+	Broker   string `json:"broker"`
+	User     string `json:"user"`
+	Password string `json:"password,omitempty"`
```
Das ist ein Bug. Das Passwort *darf* nicht raus gegeben werden!

andig commented on cmd/root.go:
```diff
@@ -157,6 +157,8 @@ func runRoot(cmd *cobra.Command, args []string) {
 	// publish to UI
 	go socketHub.Run(pipe.NewDropper(ignoreEmpty).Pipe(tee.Attach()), cache)
 
+	fmt.Println("initial publish")
```
REMOVE

andig commented on cmd/setup.go:
```diff
@@ -592,7 +592,7 @@ func configureInflux(conf globalconfig.Influx, site site.API, in <-chan util.Par
 
 // setup mqtt
-func configureMqtt(conf globalconfig.Mqtt) error {
+func configureMqtt(conf *globalconfig.Mqtt) error {
```
brauchts das? wofür?

andig committed:
Mask passwords

naltatis commented:
![Bildschirmfoto 2024-05-29 um 16 59 35](https://github.com/evcc-io/evcc/assets/152287/77d99569-abed-4e98-9117-89f48dadef65)

andig committed:
Cleanup

andig committed:
Fix deadlock

andig committed:
Always read influx config

naltatis committed:
adjusted network config model

naltatis commented:
@andig "interval" wird aktuell zwar korrekt per API in die DB geschrieben. Der Wert wird aber beim Start nicht gelesen.

naltatis commented:
Bei `residualPower` fehlt auch noch das Schreiben und Lesen in die Datenbank. Du hattest das abgehakt, aber im Code sehe ich da nichts zu. Ich hab den Haken mal wieder entfernt. 😄

naltatis committed:
adjust title modal

naltatis committed:
adjust influx modal

naltatis commented on cmd/root.go:
```diff
@@ -190,11 +186,26 @@ func runRoot(cmd *cobra.Command, args []string) {
+		}
 	}
 
+	valueChan <- util.Param{Key: keys.Influx, Val: conf.Influx}
```
@andig diese Zeile hab ich erstmal runter geschoben. Hier wurde vorher gepublisht bevor der Wert gelesen wurde. Bau gerne noch um, aber so funktionierts erstmal.

naltatis committed:
dedicated grid section; translations

naltatis committed:
readable charge status

naltatis committed:
price formatting

naltatis committed:
fixed broken yaml endpoints; experimental in code (not translation)

naltatis committed:
Merge branch 'master' into feature/config_partial_yaml

naltatis committed:
sponsor status, translation

naltatis committed:
remove startup error page

naltatis committed:
fix lint

naltatis committed:
Merge branch 'master' into feature/config_partial_yaml

andig committed:
Expose soc as temp for heating devices

andig committed:
Read/publish interval

andig commented:
Residualpower ist in der site:

	if v, err := settings.Float(keys.ResidualPower); err == nil {
		if err := site.SetResidualPower(v); err != nil {
			return err
		}
	}

andig committed:
wip

andig committed:
Add circuits yaml handler

andig committed:
Improve yaml errors

naltatis committed:
sponsor ui

naltatis commented:
Sponsor Token Handling und Statusanzeige sieht jetzt gut.

Zwei Sachen müssen noch angepasst werden:

1) Das Sponsortoken wir korrekt geschrieben/gelöscht, aber auch noch nicht in der Initialisierung verwendet bzw. migriert.
2) Der `sponsor` Key muss immer gepublisht werden. Auch wenns kein Sponsoring gibt. Sonst behält man zwischen Restarts den vorherigen State in der UI

Screenshots:

**existing sponsor**
[existing sponsor.webm](https://github.com/evcc-io/evcc/assets/152287/6eb3fd66-f11d-4eab-9f96-0309d15da800)

**trial token / expiry**
[trial token.webm](https://github.com/evcc-io/evcc/assets/152287/ad6cd2cf-7578-434f-ab66-33d1757b6ff6)

naltatis committed:
Merge branch 'master' into feature/config_partial_yaml

andig committed:
Fix residualpower

naltatis committed:
circuit default

andig committed:
Remove forced migrations and add migrate command

andig committed:
Migrate sponsor status

naltatis commented on cmd/migrate.go:
```diff
@@ -0,0 +1,62 @@
+	// TODO remove yaml file
+	settings.SetInt(keys.Interval, int64(conf.Interval))
+
+	// TODO add remaining settings @naltatis (site? loadpoints? residualpower?)
```
missing values are: `siteTitle`, `sponsorToken`, `residualPower`,  `maxGridSupplyWhileBatteryCharging`,

andig replied:
Es lassen sich nur die globalen Settings migrieren, also das Token. Der Rest ist in der global Config aktuell nicht ausmodelliert. Das müssten wir sonst noch ändern. Ich schaue nochmal...

naltatis committed:
temperature status

naltatis commented:
Formatting for temperature values.

**screenshot with fake data (since we dont visualize chargers yet)**

![Bildschirmfoto 2024-05-31 um 21 36 09](https://github.com/evcc-io/evcc/assets/152287/2b58ced8-2007-498d-803e-a222db42592f)

andig commented:
Ähhhhh- aber wieso landen wir beim Golf jetzt bei Temp statt Soc?

naltatis committed:
yaml configuration status

naltatis commented:
> Ähhhhh- aber wieso landen wir beim Golf jetzt bei Temp statt Soc?

siehe Kommentar am Screenshot 😄

naltatis commented:
Added configuration status in all boxes.

![Bildschirmfoto 2024-05-31 um 22 20 57](https://github.com/evcc-io/evcc/assets/152287/ab9896b8-a623-463c-9831-132f642b9470)

andig committed:
migrate settings

andig commented:
@naltatis statt "configured" bei Circuits the Leistungen/Ströme?

naltatis commented:
>  statt "configured" bei Circuits the Leistungen/Ströme?

Gerne. **Configured: true** ist gerade mein Save-Default der für alle YAML Entitäten geht. Da check ich einfach, ob der GET-Endpunkt Inhalt zurückliefert. Bei MQTT/Influx werden die gepublishten Details genommen. Wenn wir für die anderen Einträge auch sinnvolle Infos haben, wäre das total gut und einfach einzubauen.

Connection-Status wäre bei Influx und MQTT auch spannend. Also ob der letzte Kommunikationsversuch erfolgreich war oder nicht. Aber jetzt sind wir im Nict To Have Bereich.

naltatis committed:
Merge branch 'master' into feature/config_partial_yaml

naltatis commented on cmd/error.go:
```diff
@@ -0,0 +1,76 @@
+	ClassGo
+	ClassHEMS
+	ClassInflux
+	ClassMessenger
```
No ModbusProxy?

andig replied:
Dazu gibts ein Todo. Kann egtl. fast keinen Fehler machen ausser Port belegt. Kannnich ergänzen.

naltatis replied:
In der UI reagiere ich bereits auf `modbusproxy`. Brauchst also nur die Implementierung dafür ergänzen und dann sollte es laufen.

andig replied:
Done, Sponsortoken ebenfalls hinzu gefügt!

naltatis commented on cmd/error.go:
```diff
@@ -0,0 +1,76 @@
+	ClassCharger
+	ClassVehicle
+	ClassTariff
+	ClassCircuit
```
Error names don't match api/publishing names (circuits vs circuit, messaging vs. messager). Since mapping in the UI is done manually, this is not a problem, just wanted to make sure the singular is intentional.

andig replied:
Das wird bei Fahrzeugen ähnlich sein. To be decided…

naltatis replied:
Ja, ich würd sagen wir lassen das so. Bei Einzelgeräten sollten wir dann auch später den Identifier (db:23) mit rausgeben. Dann können wir per-device highlighten. Jetzt ists erstmal breit. Aber passt.

andig replied:
Das ist schon so 😜

naltatis committed:
error highlighting for all classes; better experimental disclaimer

naltatis commented:
Ausführlicherer Config UI Disclaimer.

![Bildschirmfoto 2024-06-02 um 12 01 31](https://github.com/evcc-io/evcc/assets/152287/ea1dcedc-95ff-44b9-a3b2-e183f59739e6)

naltatis commented:
@andig Ich hab bei `evcc migrate` noch Logausgaben hinzugefügt damit transparenter ist, was das Script tut und vor allem was auch nicht. Die Funktion ist im Disclaimer-Block (siehe Screenshot) erklärt. Für die Säuberungsfunktion habe ich da jetzt erstmal `evcc migrate --clean` angegeben. Die müsstest du noch bauen.

Achja: Migrate schreibt, zumindest bei mir. Momentan noch nichts in die Datenbank.

naltatis committed:
remove debug code

andig committed:
Add modbusproxy and sponsorship classed

andig committed:
Add migrate --reset

andig committed:
read/write database

andig commented:
All done except

> missing values are: siteTitle, sponsorToken, residualPower, maxGridSupplyWhileBatteryCharging,

andig commented:
> evcc migrate --clean

Ich habs jetzt `reset` genannt- ok?

andig commented on cmd/setup.go:
```diff
@@ -834,6 +834,7 @@ func configureTariffs(conf globalconfig.Tariffs) (*tariff.Tariffs, error) {
 }
 
 func configureDevices(conf globalconfig.All) error {
+	// TODO: add name/identifier to error for better highlighting in UI
```
@naltatis wird doch mal einen Blick in `configureMeters`. Das TODO kann weg...

andig replied:
Das war übrigens die Diskussion zum RCT vom Freitag wo wir genau das ausprobiert hatten ;)

naltatis replied:
Ok, den `device` Key hatte ich übersehen. Das funktioniert, der kommt auch an und wird jetzt im UI berücksichtigt. Allerding bekomme ich im Fehlerfall keine DB-Devices und der `/api/config/site` Endpunkt funktioniert nicht was dazu führt, dass ich keine zugeordneten Geräte anzeigen kann.

naltatis replied:
Ah, sehe gerade, das wir dafür schon Todos geschrieben und auf "Out of scope" gestellt haben. Dann ignorier das einfach. Die Fehlerkommunikation funktioniert jetzt auf jeden Fall (wenn dann die Daten kommen würden).

naltatis committed:
fixed migration values, better logging, flag shorthand (c > r)

naltatis commented:
Ich hab ein paar Anpassungen an der Migration gemacht. Nun startet es auch und kopiert die richtigen Daten.

Allerdings funktioniert `--reset` nicht. Meine Vermutung ist, dass `db.Instance.Save` nicht löscht, nur hinzufügt oder aktualisiert. Magst du noch mal schauen?

andig commented:
Wohl so ein ORM-Ding. Ich schaus an. An der Stelle müssen wir ja leider echt löschen- leer setzen reicht nicht.

andig committed:
Physically delete

andig commented on cmd/setup.go:
```diff
@@ -768,45 +819,91 @@ func configureTariffs(conf tariffConfig) (*tariff.Tariffs, error) {
+func configureModbusProxy(conf []globalconfig.ModbusProxy) error {
+	// migrate settings
+	if settings.Exists(keys.ModbusProxy) {
+		if err := settings.Yaml(keys.ModbusProxy, &conf); err != nil {
```
Was ja richtig ist- eine leere Liste von Modbus Proxies.

naltatis replied:
Jo, schlägt beim lesen trotzdem fehl. Kannst du mit der demo.yaml testen.

naltatis commented on server/db/settings/setting.go:
```diff
@@ -144,10 +185,34 @@ func Bool(key string) (bool, error) {
+		return ErrNotFound
+	}
+
+	other := make(map[string]any)
```
@andig das hier scheint der Grund für das ModbusProxy Leseproblem zu sein. Im Gegensatz zu den meisten anderen Yamls (circuits auch?) haben wir hier keine Map.

naltatis replied:
Der Fehler hier existiert immer noch. Heißt nach `migrate` startet evcc nicht mehr.

naltatis committed:
dark mode fix

naltatis commented:
@andig die `--reset` Änderung funktioniert bei mir noch nicht. Ich bekomme folgendes:

```
go run main.go --config cmd/demo.yaml migrate --reset

[main  ] INFO 2024/06/05 09:09:10 evcc 0.0.0
[main  ] INFO 2024/06/05 09:09:10 using config file: cmd/demo.yaml
[db    ] INFO 2024/06/05 09:09:10 using sqlite database: /Users/michael/.evcc/evcc.db
[main  ] WARN 2024/06/05 09:09:10 resetting:
[main  ] DEBUG 2024/06/05 09:09:10 - global settings
panic: runtime error: slice bounds out of range [15:1:]

goroutine 1 [running]:
slices.Delete[...](...)
	/opt/homebrew/Cellar/go/1.22.1/libexec/src/slices/slices.go:219
github.com/evcc-io/evcc/server/db/settings.Delete({0x102b30b09, 0x8})
	/Users/michael/lab/evcc/server/db/settings/setting.go:81 +0x324
github.com/evcc-io/evcc/cmd.runMigrate(0x1067a3900, {0x14000a8bdd0?, 0x4?, 0x102b207e9?})
	/Users/michael/lab/evcc/cmd/migrate.go:44 +0x1f0
github.com/spf13/cobra.(*Command).execute(0x1067a3900, {0x1400112f780, 0x3, 0x4})
	/Users/michael/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:987 +0x828
github.com/spf13/cobra.(*Command).ExecuteC(0x1067a52e0)
	/Users/michael/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1115 +0x344
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/michael/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1039
github.com/evcc-io/evcc/cmd.Execute()
	/Users/michael/lab/evcc/cmd/root.go:103 +0x24
main.main()
	/Users/michael/lab/evcc/main.go:44 +0x38
exit status 2
```

naltatis committed:
Merge branch 'master' into feature/config_partial_yaml

naltatis committed:
device specific errors

naltatis commented:
Ich hab `dirty` via Websocket mal auf "later" verschoben. Das ist ein Thema was wir schön isoliert angehen können und der jetzige Stand funktioniert ja.

naltatis committed:
add e2e tests for tariffs yaml edit

naltatis committed:
test invalid yaml; debug

naltatis committed:
debug: more logging

naltatis committed:
cross-platform key modifier

naltatis committed:
added mqtt e2e tests (json)

naltatis committed:
show startup error message on main screen; add red badge if fatal

naltatis commented:
"Startup Error" Nachrichtauf Hauptscreen + Red Badge in Navigation

![Bildschirmfoto 2024-06-07 um 09 45 19](https://github.com/evcc-io/evcc/assets/152287/8034e316-fe13-44b5-8fef-5e0d24f932cd)

naltatis commented:
@andig meine Todoliste hier ist jetzt leer. Wenn die Fehler aus dem Migrationscommand behoben sind können wir den mMn. reinnehmen.

andig committed:
Fix panic

andig commented:
Panic auch behoben. Man muss es halt auch richtig machen :) Wir könnten das jetzt als `:testing` Docker releasen?

naltatis commented:
... oder mutig zu nehm Nightly machen. Wir haben die Änderungen ja so umgebaut, dass eigentlich nichts Breaking sein sollt

naltatis commented:
> Panic auch behoben.

Confirm. Löschen klappt jetzt wie erwartet.

andig committed:
wip

andig committed:
Honor mapstructure target object kind

andig committed:
wip

naltatis committed:
added circuits to migration

andig committed:
wip

andig committed:
Merge branch 'master' into feature/config_partial_yaml

naltatis commented:
🎊
